Server-errors-and-solutions.md

```diff
@@ -0,0 +1,80 @@
+
+### Preventive action
+
+* __Manual testing__: When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
+Testing with foreign / special characters (as mentioned under Template Users).
+
+* __Code__: Audit jobs need to account for all instances or occurrences of the relevant item on the server. Sufficient written proof / explanation about the “thoroughness” of the audit can be added as a comment in the audit job code.
+
+* __Review phase__: Reviewers should be encouraged to ask questions and ensure that the audit job is thorough.
+
+## Race Conditions / Slow Network Issues
+
+Some errors happen when interacting too fast with the application or when the network is too slow. Example: Clicking on the “content” tab in the translation editor before it has fully loaded (#6378). Other e.g. errors #6161, #13087, #8181, #8733, #8321, #7381, #7252, #7086, #7041, #6921, #6298, #6161   
+
+### Preventive action
+
+* __Manual testing__: For code that has a lot of “async” functionality, the authors should be encouraged to try throttling their network and/or interacting with the UI before it has completely loaded. This could be a category of “template users” mentioned here.
+* __Automated testing__: Configure certain automated tests to run under throttled network conditions. Might be worth considering the introduction of tests where the test interacts too quickly with elements on the page and expect nothing to happen (there should be no console errors).
+
+## Related to a migration
+
+One-off errors from migration of libraries or frameworks e.g. missing imports in the module files.
+
+### Preventive action
+* __Documentation__: Record common issues in the relevant migration page when they are identified.
+
+## Errors that didn't need to be logged
+Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors. Other e.g. errors: [$compile:tpload]" issue causing page load errors #5793, #6500, #8761, #7563 
+
+### Preventive action
+
+* __Code / review__: When adding logging statements or throwing errors that are not handled, always consider what scenarios and how often they will get called. If the frequency is high, discuss with the server error coordinator for the month to check if there are alternatives to achieve the same (e.g. it may be possible to add logs on stackdriver interface directly in a controlled way.)
```

Comment:
what scenarios --> what scenarios these arise in

same --> desired logging behaviour